### DISCUSSION TITLE
Why ArrayBuffer?

### DISCUSSION THREAD

**Discussion by bakkot** [humanwhocodes/humanfs#12] [Feedback]
I liked your [blog post](https://humanwhocodes.com/blog/2024/01/fsx-modern-filesystem-api-javascript/), but one thing caught my eye: `fsx.arrayBuffer(filePath)`.

ArrayBuffer is almost never what you want because you can't read from one directly. You have to create a view like a Uint8Array to actually read it. And [lots](https://nodejs.org/api/http.html#requestenddata-encoding-callback) [of](https://nodejs.org/api/child_process.html#child_processexecsynccommand-options) [methods](https://nodejs.org/api/crypto.html#cipherupdatedata-inputencoding-outputencoding) don't accept an ArrayBuffer, only a Uint8Array or Buffer wrapper.

That seems counter to the ethos of "Common cases should be easy". The common case for reading bytes wants a Uint8Array, not an ArrayBuffer.

And of course if you _do_ need an ArrayBuffer for some reason (though this is _extremely_ rare) you can do `bytes().buffer` and get it.

**Comment by nzakas** [comment]
That's a fair assessment. I initially started out by copying methods from the [`Response` object](https://developer.mozilla.org/en-US/docs/Web/API/Response/arrayBuffer), assuming that's what people would be familiar with and want (along with `text()` and `json()`), but perhaps this one is worth revisiting.

**Reply by bakkot** [reply]
Yeah, I think that's kind of a legacy thing. At this point the web platform [generally uses Uint8Array for byte buffers](https://github.com/whatwg/fetch/issues/1724#issuecomment-1814626681).

**Reply by bakkot** [reply]
(Incidentally if you were going to provide a method to vend a `Uint8Array`, I would just call it `.bytes()`.)

**Reply by bakkot** [reply]
Yeah, [looks like it is probably a legacy thing](https://matrixlogs.bakkot.com/WHATWG/2024-01-18#L0).

**Comment by nzakas** [comment]
The new `bytes()` method has been added and `write()` now access ArrayBuffer view as an argument.

**Reply by bakkot** [reply]
Nice, though checking the implementation for `write`, it looks like [you're writing the entire backing `.buffer` to the output](https://github.com/humanwhocodes/fsx/blob/1fd55174a528ef3dcbabc154347006bec799f3f9/packages/node/src/node-fsx.js#L198), instead of only the part actually viewed by the Uint8Array.

To give a concrete example:

```js
let base = new Uint8Array([5, 6, 7, 8]);
let sub = base.subarray(1, 3);
await fsx.write("file", sub);
```
should only write the fours bytes `6` and `7` to "file", not all four bytes.

**Reply by nzakas** [reply]
Please open an issue.
